src/malli/util.cljc

```diff
@@ -34,6 +35,128 @@
 ;; public api
 ;;
 
+
+;; pre+post walking with state map
+
+(defn walk*
+  "Prewalk recursively over the Schema with inner and its children then postwalk with outer.
+  The inner (prewalk) callback is a arity-3 function. It takes schema, path, and options.
+  It returns a vector of [schema options]--the new options are passed to children prewalks.
+  The outer (postwalk) callback is a arity4 function with the following
+  arguments: schema, path, (walked) children, and options. By default, returns its schema
+  argument."
+  ([?schema inner]
+   (walk* ?schema inner nil))
+  ([?schema inner options]
+   (walk* ?schema inner (fn [s _p _c _options] s) options))
+  ([?schema inner outer options]
+   (let [[s options] (inner (m/schema ?schema options) [] options)]
+     (m/-walk
+       s
+       (reify m/Walker
+         (-accept [_ s _ _] s)
+         (-inner [this s p options] (let [[s options] (inner s p options)]
+                                      (m/-walk s this p options)))
+         (-outer [_ s p c options] (outer (m/-set-children s c) p c options)))
+       [] options))))
+
+;; free variables
+
+(defn schema->fvs 
+  "Returns the free variables in the schema."
+  [schema]
+  (let [fvs-atom (atom #{})
+        rec! (fn rec! [schema ref-scope]
+               (walk* schema
+                      (fn [schema _path options]
+                        (let [registry (-> schema m/properties :registry)
+                              registry-rhs-ref-scope (reduce disj (::ref-scope options) (keys registry))
+                              _ (run! (fn [[k s]]
+                                        (rec! s registry-rhs-ref-scope))
+                                      registry)
+                              new-ref-scope (into (::ref-scope options) (keys registry))
+                              is-var-reference? (satisfies? m/RefSchema schema)
+                              is-not-in-scope? (not (contains? new-ref-scope (m/-ref schema)))
+                              _ (when (and is-var-reference?
+                                           is-not-in-scope?)
+                                  (swap! fvs-atom conj (m/-ref schema)))]
+                          [schema (c/assoc options ::ref-scope new-ref-scope)]))
+                      {::ref-scope ref-scope})
+               nil)]
+    (rec! schema #{})
+    @fvs-atom))
+
+(declare subst-schema)
+
+(comment
+  (alpha-rename-schema [:schema {:registry {::foo :int}}
+                        ::foo]
+                       {})
+  ((requiring-resolve 'clojure.repl/pst) 100)
+  )
+
+(defn alpha-rename-schema [schema options]
```

Comment:
@ikitommi I need some help creating a function that performs alpha renaming on schemas.

I seem to be able to rename the variables just fine, but on `m/deref` something funny is happening--I'm guessing the ref renames aren't being fully propagated.

```clojure
  (is (= :int
         (m/form
           (m/deref-all
             (mu/alpha-rename-schema [:schema {:registry {::foo :int}}
                                      ::foo]
                                     {})))))
; (err) NullPointerException Cannot invoke "clojure.lang.IFn.invoke()" because "this._ref" is null
; (err) 	malli.core/-ref-schema/reify/reify--5822 (core.cljc:1335)
; (err) 	malli.core/deref (core.cljc:1899)
; (err) 	malli.core/deref (core.cljc:1893)
; (err) 	malli.core/deref-all (core.cljc:1906)
; (err) 	malli.core/deref-all (core.cljc:1901)
; (err) 	malli.core/deref-all (core.cljc:1904)
; (err) 	malli.core/deref-all (core.cljc:1901)
; (err) 	malli.util-test/eval18748 (NO_SOURCE_FILE:1181)
; (err) 	malli.util-test/eval18748 (NO_SOURCE_FILE:1179)
; (err) 	clojure.lang.Compiler.eval (Compiler.java:7181)
; (err) 	clojure.lang.Compiler.eval (Compiler.java:7136)
; (err) 	clojure.core/eval (core.clj:3202)
; (err) 	clojure.core/eval (core.clj:3198)
; (err) 	nrepl.middleware.interruptible-eval/evaluate/fn--963/fn--964 (interruptible_eval.clj:87)
; (err) 	clojure.core/apply (core.clj:667)
; (err) 	clojure.core/with-bindings* (core.clj:1977)
; (err) 	clojure.core/with-bindings* (core.clj:1977)
; (err) 	nrepl.middleware.interruptible-eval/evaluate/fn--963 (interruptible_eval.clj:87)
; (err) 	clojure.main/repl/read-eval-print--9110/fn--9113 (main.clj:437)
; (err) 	clojure.main/repl/read-eval-print--9110 (main.clj:437)
; (err) 	clojure.main/repl/fn--9119 (main.clj:458)
; (err) 	clojure.main/repl (main.clj:458)
; (err) 	clojure.main/repl (main.clj:368)
; (err) 	nrepl.middleware.interruptible-eval/evaluate (interruptible_eval.clj:84)
; (err) 	nrepl.middleware.interruptible-eval/evaluate (interruptible_eval.clj:56)
; (err) 	nrepl.middleware.interruptible-eval/interruptible-eval/fn--994/fn--998 (interruptible_eval.clj:152)
; (err) 	nrepl.middleware.session/session-exec/main-loop--1062/fn--1066 (session.clj:202)
; (err) 	nrepl.middleware.session/session-exec/main-loop--1062 (session.clj:201)
; (err) 	java.lang.Thread.run (Thread.java:831)
```

How would you write this function? See `malli.util-test/alpha-rename-schema-test` for what I want to achieve.

Reply:
I'll check this.